pkg/cloud/services/eks/cluster.go

```
@@ -141,6 +141,10 @@ func (s *Service) reconcileCluster(ctx context.Context) error {
 }
 
 func (s *Service) setStatus(cluster *eks.Cluster) error {
+	// Set the current Kubernetes control plane version in the status.
+	s.scope.ControlPlane.Status.Version = cluster.Version
```

We may need to do some manipulation here. AWS will return something like `1.31` and not a full semver.

The Version field in the spec is in the format of vX.Y.Z or vX.Y

So if AWS returns `1.31` we should put `v1.31` or `v1.31.0 into the field

---

Hm. That's a bit tricky then.
What if the user had `v1.31.4` and then specifies a `v1.31.5`, to upgrade, for example?
How do we know if that was successfully rolled out if we can only read `1.31`?

---

I had a deeper look at this. From what I have been able to gather AWS EKS directly manages patch versions and so the user should only specify `major.minor` in the `Version` field of the SDK/Console, even though that's not explicitly specified in the API godoc (https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#CreateClusterInput).

It seems though that if you provide a version with a patch (for example, 1.30.2), EKS will ignore the patch portion and effectively use the major.minor version (1.30). It does not reject the request; it silently truncates the patch version.

This behavior occurs because EKS manages Kubernetes patch updates automatically and only supports specifying the version in a major.minor format.

---

I also had a look at the [controlPlaneStablePreflightCheck](https://github.com/kubernetes-sigs/cluster-api/blob/3abcd216748c1dcae6828afb317099526e237ae9/internal/controllers/machineset/machineset_preflight.go#L94) where the strictest check on spec/status.Version is the [IsUpgrading](https://github.com/kubernetes-sigs/cluster-api/blob/3abcd216748c1dcae6828afb317099526e237ae9/internal/contract/controlplane.go#L210-L241) one.

In this check [`semver.ParseTolerant()`](https://github.com/blang/semver/blob/2ee87856327ba09384cabd113bc6b5d174e9ec0f/semver.go#L203-L224) is used for both spec.Version and status.Version, which effectively trims the `v` and automatically puts a trailing `.0` patch section on the version if not specified.

This is then followed up by a [`version.Compare(spec.Version, status.Version, ...)`](https://github.com/kubernetes-sigs/cluster-api/blob/3abcd216748c1dcae6828afb317099526e237ae9/util/version/version.go#L99-L104) which marks IsUpgrading=true when spec.Version is greater than status.Version.

---

So considering:
- `s.scope.ControlPlane.Status.Version = cluster.Version` sets `status.Version` to the `major.minor` current version returned by the AWS EKS API (e.g. `1.31`) 
- considering what said in (b), that the `v`s are ignored, and semver normalised
- considering that what is specified in `spec.Version` is then internally reduced by the EKS API to `major.minor`

I see two viable options:

option 1) change the `spec.Version` validation https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/8cc4bebb0737a89dcd05fcb4e64673ef8db13d6f/controlplane/eks/api/v1beta2/awsmanagedcontrolplane_types.go#L72-L74 to restrict it to only `major.minor`. This should then always result in a comparable version between what the user provides (spec.Version) and what AWS EKS spits out (status.Version). This is likely a validation breaking change.

option 2) rather than directly setting `s.scope.ControlPlane.Status.Version = cluster.Version`, we cheat and look at what the user set in the spec.Version (`s.scope.ControlPlane.Spec.Version`), we then look at whether they set a patch version, we check it matches with the `major.minor` returned by AWS EKS API, and if it does we set the exact/or a greater `major.minor.patch` in status.Version to what the user specified. In case we opt for a greater one, we could use `major.minor.999` or something that highlights it is always greater.

WDYT?

---

@damdo - great write up and thanks for looking into [semver.ParseTolerant()](https://github.com/blang/semver/blob/2ee87856327ba09384cabd113bc6b5d174e9ec0f/semver.go#L203-L224) and proposing options.

Option 1) - i think this is the better long term solution, but it will require an API version bump as we are changing the validation rules on the field. We'd need to have the conversion functions offered by the API version bump to be able to change the value of the version field, from something like `v1.31.2` to `v1.31`, for any existing AWSManagedControlPlanes.

Option 2 ) - is the easier short term solution.

---

Updated the PR with Option 2.

---

Added option 1 to the issue tracking v1beta3 changes we want to make

---

Great! Thanks!
I've fixed a nil pointer exception that was happening in the mock tests.
So this is done too now.
PR ready for a review @richardcase